src/cpp/rtps/builtin/discovery/database/DiscoveryDataBase.cpp

```diff
@@ -241,6 +280,240 @@ void DiscoveryDataBase::AckedFunctor::operator () (
     pending_ |= !is_acked;
 }
 
+void DiscoveryDataBase::process_dispose_writer(
+        eprosima::fastrtps::rtps::CacheChange_t* ch)
+{
+    const eprosima::fastrtps::rtps::GUID_t& writer_guid = guid_from_change(ch);
+
+    auto wit = writers_.find(writer_guid);
+    if (wit != writers_.end())
+    {
+        wit->second.set_disposal(ch);
+    }
+
+    auto pit = participants_.find(writer_guid.guidPrefix);
+    if (pit != participants_.end())
+    {
+        pit->second.remove_writer(writer_guid);
+    }
+
+    for (auto tit = writers_by_topic_.begin(); tit != writers_by_topic_.end(); ++tit)
+    {
+        tit->second.erase(
+                std::remove(tit->second.begin(), tit->second.end(), writer_guid),
+                tit->second.end());
+    }
```

Comment:
Can't we get the topic from the writer and then only iterate on that topic looking for the writer to remove it. Once you find it, you could break the loop, cause it will only be there once.